internal/controller/pipelinerollout/pipelinerollout_progressive.go

```diff
@@ -103,17 +104,30 @@ func (r *PipelineRolloutReconciler) AssessUpgradingChild(
 		return areAllVerticesReplicasReady, replicasFailureReason, nil
 	}
 
+	// if Assessment EndTime ever arrives, we automatically fail
+	if childStatus.AssessmentEndTimeArrived() {
+		numaLogger.WithValues("childStatus", *childStatus).Debug("upgrading child EndTime has arrived, failing the assessment")
+		// set AssessmentEndTime to now and return failure
+		_ = progressive.UpdateUpgradingChildStatus(pipelineRollout, func(status *apiv1.UpgradingChildStatus) {
+			currentTime := metav1.NewTime(time.Now())
+			status.BasicAssessmentEndTime = &currentTime
+		})
+
+		return apiv1.AssessmentResultFailure, "Assessment EndTime has arrived", nil
+	}
+
 	// First perform basic resource health check
 	assessment, reasonFailure, err := progressive.PerformResourceHealthCheckForPipelineType(ctx, existingUpgradingChildDef, verifyReplicasFunc)
 	if err != nil {
 		return assessment, reasonFailure, err
 	}
-	// if we fail even once, that's considered failure
+	// if we fail even once, that's considered a failure
 	if assessment == apiv1.AssessmentResultFailure {
 		// set AssessmentEndTime to now and return failure
 		_ = progressive.UpdateUpgradingChildStatus(pipelineRollout, func(status *apiv1.UpgradingChildStatus) {
-			assessmentEndTime := metav1.NewTime(time.Now())
-			status.BasicAssessmentEndTime = &assessmentEndTime
+			currentTime := metav1.NewTime(time.Now())
+			status.BasicAssessmentEndTime = &currentTime
+			status.AssessmentLastUpdated = &currentTime
 		})
```

Comment:
No, with this change, now we are allowed to fail. We just need one consecutive minute of success.

Take this example:
Say PipelineRollout has this schedule:
        - kind: Pipeline
          schedule: "120,360,60,10"
This means we start checking after 120 seconds after pipeline created, we continue checking until 360 seconds after pipeline created, we make sure it's healthy for a consecutive window of 60 seconds.

Say this Pipeline is not healthy between 120 seconds and 180 seconds. But say it is healthy between 150 seconds and 210 seconds. We want it to be assessed healthy in this case.